Fix #77 Add API controller with token_authenticate_user method

mrjoshua22 opened the pull request:
Добавлен метод token_authenticate_user, тесты к нему и гем 'rails-controller-testing'. Метод тестируется через создание анонимного контроллера и использовании в нем метода через before_action. Для проверки значения возвращаемой инстанс переменной @current_user добавлен гем 'rails-controller-testing'. Тесты написаны для запроса с корректным токеном, токена с некорректным payload и запроса без токена вообще.

mrjoshua22 commented:
Если @current_user == nil, то нужно возвращать status :unauthorized. Добавлю следующим коммитом.

mrjoshua22 commented:
Добавлена обработка ActiveRecord::RecordNotFound и JWT::DecodeError

XaoGao commented on Gemfile:
```diff
@@ -42,6 +42,7 @@ end
 
 group :test do
   gem 'capybara', '>= 3.26'
+  gem 'rails-controller-testing'
```
вот что-то мне ненравиться что добавлен этот гем, тем более стоит зафикисировать гем не определенной версии

mrjoshua22 replied:
Гем убрал

XaoGao commented on app/controllers/api/api_controller.rb:
```diff
@@ -1,3 +1,18 @@
+  class ApiController < ApplicationController
+    protect_from_forgery with: :null_session
+
+    def token_authenticate_user
```
Вообще конечно стоит это вынести в отдельный controllers/concerns, но пока можно оставить и тут.
Некорректно с наименованием **request.headers["Authorization"]** - это не token а header. Тем более не стоит переопределять значенеи перменной, а написать явно
```ruby
header = request.headers["Authorization"]
token = header.split.last if header
```
Так же стоит явно указать что если 
```ruby
if !token
  return unauthorized
end
```
Добавить проверку что token.split.first - это слово Bearer - так как могу идти запросы с другой схемой.
Не нужно явно указывать **I18n**, можно просто вызвать метод **t**

mrjoshua22 replied:
Переопределение переменной убрал. 
Добавил проверку header["Athorization"] на nil и другой тип токена (не Bearer) в одном условии через ||. 
I18n из контроллера убрал, но оставил в тестах - иначе undefined_method 't' выбрасывает. 
По разным статусам в каждом из условий - просмотрел все статусы на 4хх - ни одного подходящего не нашел. Изменил только сообщение об ошибке в случае ошибки ActiveRecord::RecordNotFound.

XaoGao commented on spec/controllers/api/api_controller_spec.rb:
```diff
@@ -0,0 +1,54 @@
+require "rails_helper"
+
+RSpec.describe Api::ApiController do
+  subject { described_class }
```
Ощущение что это тут бесполезно

mrjoshua22 replied:
Убрал.

XaoGao commented on spec/controllers/api/api_controller_spec.rb:
```diff
@@ -0,0 +1,54 @@
+    it "bearer token is valid" do
+      request.headers["Authorization"] = "Bearer #{token}"
+      get :fake_action
+      current_user = assigns(:current_user)
```
это очень плохой знак, что мы для проверки пытаемся получить инстансную переменную объекта, тесты нужны чтобы проверить не структуру а логику, мое предложение такие, в классе Api::ApiController создать метод current_user который вернет @current_user и тогда мы можем обращется к нему в тесте 
```
controller.current_user
```

mrjoshua22 replied:
Переписал с использованием дополнительного метода контроллера current_user.

XaoGao commented:
@mrjoshua22 указал замечания

XaoGao commented:
Вот хороший способ избавиться от JSON.parse(response.body) https://stackoverflow.com/questions/5159882/how-to-check-for-a-json-response-using-rspec
```ruby
module ApiHelpers
  def json_body
    JSON.parse(response.body)
  end
end

RSpec.configure do |config| 
  config.include ApiHelpers, type: :request
end
```

XaoGao commented:
@mrjoshua22 все выгрлядит правильно, но можно еще немного улучшить код
```ruby
 def token_authenticate_user
  # В будущем будут другие статусы ответа
  if authorization_header.blank?
    render status: :unauthorized, json: { error: t("api.errors.unauthorized") } and return
  end

  if authenticate_scheme_bearer?
    render status: :unauthorized, json: { error: t("Wrong scheme, you have to use a Bearer") } and return
  end

  if token.nil? || token.empty?
    render status: :unauthorized, json: { error: t("Token is nil or empty") } and return
  end

  if token_expired?
    render status: :unauthorized, json: { error: t("Token is expired") } and return
  end

  begin
    set_user
  rescue ActiveRecord::RecordNotFound
    render status: :unauthorized, json: { error: t("api.errors.not_found") }
  rescue JWT::DecodeError
    render status: :unauthorized, json: { error: t("unexpected error") }
  end
end

private

# приватные методы не тестируются

def authorization_header
  request.headers["Authorization"]
end

def token
  authorization_header.split.last
end

def authenticate_scheme_bearer?
  authenticate_scheme != "Bearer"
end


def authenticate_scheme
  authorization_header.split.first 
end

def decoded
  JwtService.decode(token)
end

def set_user
  @current_user ||= User.find(decoded[:id])
end

# эту проверку забыл указать в задаче
def token_expired?
  DateTime.now.to_i > decoded[:exp]
end
```

mrjoshua22 commented:
Вопросы по замечаниям:
1. Почему именно проверка токена идет через token.nil? || token.empty? если можно сделать token.blank?
2. Название метода authenticate_scheme_bearer? несколько сбивает с толку. Из названия кажется, что он должен возвращать true если header.split.first == 'Bearer', а он наоборот возвращает true при  header.split.first != 'Bearer'.
Может стоит назвать что-то вроде authenticate_scheme_wrong?
3. Все переменные определяются внутри приватных методов. Как определить когда стоит напрямую присваивать значения переменным, как в JwtService или SessionsController, а когда стоит все построить на приватных методах?

XaoGao commented:
@mrjoshua22 
1) Можно использовать **blank?**
2) authenticate_scheme_wrong - мне тут не нравиться **wrong** малоинформативно, давай просто сделаем
```ruby
unless authenticate_scheme_bearer?
  render status: :unauthorized, json: { error: t("Wrong scheme, you have to use a Bearer") } and return
end
def authenticate_scheme_bearer?
  authenticate_scheme == "Bearer"
end
```
3) Нет четких критерии, нужно попробовать написать в обоих стилях и посмотреть что лучше читается

mrjoshua22 commented:
По последним замечаниям:
1. Метод token.blank? (nil?, empty?) использовать нельзя, так как token возвращает последний элемент массива authorization_header, соответственно в случае пустого токена будет возвращаться Bearer, и проверки не проходят. Для проверки добавил метод token_blank? где сравнивается token и authenticate_scheme - последний и первый элемент массива. Можно было прописать через authorization_header.size == 1, но сделал вот так.
По этой же причине в тестах прописан только вариант token blank - на token nil нет смысла писать, получится одно и то же.
2.  Проверку token_expired? убрал вообще вместе с приватным методом, так как JWT сам проверяет срок действия токена и возвращает JWT::ExpiredSignature в случае просрочки. Поэтому добавил еще один блок rescue, иначе токен не может быть декодирован в случае истечения срока действия.
3. Оказывается метод current_user изначально доступен благодаря Devise через ApplicationController, поэтому отдельно прописывать его для проверки значения @current_user в тестах не нужно.
4. В api_helper добавил type: :controller после config.include ApiHelpers, потому что так правильней.
